src/placeos-frontend-loader/api/remote/gitlab.cr

```diff
@@ -0,0 +1,99 @@
+require "hash_file"
+require "./remote"
+require "gitlab"
+
+module PlaceOS::FrontendLoader
+  class GitLab < PlaceOS::FrontendLoader::Remote
+    def initialize(@metadata : Metadata = Metadata.instance)
+    end
+
+    private alias Remote = PlaceOS::FrontendLoader::Remote
+
+    ENDPOINT = "https://gitlab.com/api/v4"
+
+    @gitlab_client = Gitlab.client(ENDPOINT, GITLAB_TOKEN)
+
+    def get_repo_id(repo_name : String)
+      repo = URI.encode_www_form(repo_name)
+      @gitlab_client.project(repo)["id"].to_s.to_i
+    end
+
+    # Returns the branches for a given repo
+    def branches(repo : String) : Hash(String, String)
+      repo_id = get_repo_id(repo)
+      @gitlab_client.branches(repo_id).as_a.each_with_object({} of String => String) do |branch, branches|
+        branch_name = branch["name"].to_s
+        branches[branch_name] = branch["commit"]["id"].to_s
+      end
+    end
+
+    # Returns the commits for a given repo on specified branch
+    def commits(repo : String, branch : String = "master") : Array(Remote::Commit)
+      repo_id = get_repo_id(repo)
+      @gitlab_client.commits(repo_id).as_a.map do |comm|
+        Remote::Commit.new(
+          commit: comm["id"].as_s,
+          date: comm["authored_date"].as_s,
+          author: comm["author_email"].as_s,
+          subject: comm["title"].as_s
+        )
+      end
+    end
+
+    # Returns the release tags for a given repo
+    def releases(repo : String) : Array(String)
+      repo_id = get_repo_id(repo)
+      @gitlab_client.tags(repo_id).as_a.map do |value|
+        value["name"].to_s
+      end
+    end
+
+    def url(repo_name : String) : String
+      "https://gitlab.com/#{repo_name}"
+    end
+
+    def download(
+      ref : Remote::Reference,
+      path : String,
+      branch : String? = "master",
+      hash : String? = "HEAD",
+      tag : String? = nil
+    )
+      repository_uri = url(ref.repo_name)
+      repository_folder_name = path.split("/").last
+
+      hash = get_hash(hash, repository_uri, tag, branch)
+      temp_tar_name = Random.rand(UInt32).to_s
+
+      Git.repository_lock(repository_folder_name).write do
+        Log.info { {
+          message:    "downloading repository",
+          repository: repository_folder_name,
+          branch:     branch,
+          uri:        repository_uri,
+        } }
+
+        begin
+          repo_encoded = ref.repo_name.gsub("/", "%2F")
+          archive_url = "https://gitlab.com/api/v4/projects/#{repo_encoded}/repository/archive.tar.gz?sha=#{hash}"
+
+          download_archive(archive_url, temp_tar_name)
```

Comment:
This doesn't use the file handle returned by `download_archive`.
I think this could potentially be leaking open files

Reply:
`download_archive` doesn't return anything it just saves to a path that `extract_archive` looks at